src/target/riscv/riscv.c

```diff
@@ -604,90 +605,146 @@ static int maybe_add_trigger_t1(struct target *target, struct trigger *trigger)
 	return ERROR_OK;
 }
 
-static int maybe_add_trigger_t2(struct target *target, struct trigger *trigger)
+struct trigger_request_info {
+	riscv_reg_t tdata1;
+	riscv_reg_t tdata2;
+	riscv_reg_t imask;
+};
+
+static int try_setup_single_match_trigger(struct target *target,
+		struct trigger *trigger, int trigger_type,
+		struct trigger_request_info trig_info)
 {
-	int ret;
-	riscv_reg_t tdata1, tdata2;
+	int ret = ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
+	RISCV_INFO(r);
 
+	for (unsigned int idx = 0;
+			find_next_trigger(target, trigger_type, false, &idx) == ERROR_OK &&
+			ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
+			++idx) {
+		ret = set_trigger(target, idx, trig_info.tdata1, trig_info.tdata2,
+				trig_info.imask);
+		if (ret == ERROR_OK) {
+			r->trigger_unique_id[idx] = trigger->unique_id;
+			return ERROR_OK;
+		}
+	}
+	return ret;
+}
+
+static int try_setup_chained_match_triggers(struct target *target,
+		struct trigger *trigger, int trigger_type,
+		struct trigger_request_info t1,
+		struct trigger_request_info t2)
+{
+	int ret = ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
 	RISCV_INFO(r);
 
-	tdata1 = 0;
+	for (unsigned int idx = 0;
+			find_next_trigger(target, trigger_type, true, &idx) == ERROR_OK &&
+			ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
+			++idx) {
+		ret = set_trigger(target, idx, t1.tdata1, t1.tdata2, t1.imask);
+		if (ret != ERROR_OK)
+			continue;
+		ret = set_trigger(target, idx + 1, t2.tdata1, t2.tdata2, t2.imask);
+		if (ret == ERROR_OK) {
+			r->trigger_unique_id[idx] = trigger->unique_id;
+			r->trigger_unique_id[idx + 1] = trigger->unique_id;
+			return ERROR_OK;
+		} else
+			/* Undo the setting of the previous trigger */
+			set_trigger(target, idx, 0, 0, 0);
+	}
+	return ret;
+}
+
+static int setup_match_trigger(struct target *target,
+		struct trigger *trigger, int trigger_type,
+		struct trigger_request_info t_napot, struct trigger_request_info t1_ge,
+		struct trigger_request_info t2_lt, struct trigger_request_info t1_lt,
+		struct trigger_request_info t2_ge, struct trigger_request_info t_eq)
+{
+	int ret = ERROR_OK;
+
+	if (!(trigger->execute || trigger->length == 1)) {
```

Comment:
```suggestion
	if (!trigger->execute && trigger->length != 1) {
```

Reply:
(Or even more idiomatic: `trigger->length > 1`)

Anyway, I'd recommend to move the control flow code (sequence of actions attempted) into the upper-level function, as discussed [in another comment](https://github.com/riscv/riscv-openocd/pull/800#discussion_r1109459134)

Reply:
Addressed.